### ISSUE TITLE
Canary Rollout Fails to Update `replicas` and `readyReplicas`

### ISSUE THREAD

**Issue by gavin-db** [argoproj/argo-rollouts#1860]
## Summary 

During a release of a service `ngrok-proxy` using a canary strategy (no HPA or traffic shifting), `replicas` and `readyReplicas` are not correctly decremented when the last stable pod is deleted, resulting in the rollout [stalling indefinitely](https://github.com/argoproj/argo-rollouts/blob/85cdf40cbd1e4e4b11d1515e6991f5a664b41669/utils/rollout/rolloututil.go#L129) with `old replicas are pending termination`, even though all old replicas have been terminated.

When the last pod is deleted, the controller logs begin to repeatedly display a reconciliation error, even though reconciliation should be complete as the stable pods are fully deleted (this was verified from kubernetes event logs and is reflected in the messages posted in the Rollout status). The error involves `spec.securityContext.hostNetwork` host network port rules, although these errors do not show up anywhere else (including event logs) and do not prevent any pods from being scheduled or deleted, and therefore should not affect reconciliation.

I can consistently reproduce this issue.

## Diagnostics

What version of Argo Rollouts are you running?
v1.1
# Controller Logs
**These logs only become interesting after the last pod is deleted:**
```
`time="2022-02-10T04:29:08Z" level=info msg="Reconciliation completed" generation=2 namespace=development resourceVersion=943663507 rollout=ngrok-proxy time_ms=1148.469979
time="2022-02-10T04:29:08Z" level=info msg="Started syncing rollout" generation=2 namespace=development resourceVersion=943663544 rollout=ngrok-proxy
time="2022-02-10T04:29:08Z" level=info msg="Patched conditions: {\"status\":{\"conditions\":[{\"lastTransitionTime\":\"2022-02-10T03:31:48Z\",\"lastUpdateTime\":\"2022-02-10T03:31:48Z\",\"message\":\"Rollout has minimum availability\",\"reason\":\"AvailableReason\",\"status\":\"True\",\"type\":\"Available\"},{\"lastTransitionTime\":\"2022-02-10T03:56:01Z\",\"lastUpdateTime\":\"2022-02-10T03:56:01Z\",\"message\":\"RolloutCompleted\",\"reason\":\"RolloutCompleted\",\"status\":\"False\",\"type\":\"Completed\"},{\"lastTransitionTime\":\"2022-02-10T04:29:07Z\",\"lastUpdateTime\":\"2022-02-10T04:29:08Z\",\"message\":\"ReplicaSet \\\"ngrok-proxy-69b849644c\\\" is progressing.\",\"reason\":\"ReplicaSetUpdated\",\"status\":\"True\",\"type\":\"Progressing\"},{\"lastTransitionTime\":\"2022-02-10T04:29:08Z\",\"lastUpdateTime\":\"2022-02-10T04:29:08Z\",\"message\":\"Rollout is paused\",\"reason\":\"RolloutPaused\",\"status\":\"False\",\"type\":\"Paused\"}]}}" generation=2 namespace=development resourceVersion=943663544 rollout=ngrok-proxy
time="2022-02-10T04:29:09Z" level=info msg="persisted to informer" generation=2 namespace=development resourceVersion=943663554 rollout=ngrok-proxy
time="2022-02-10T04:29:09Z" level=error msg="roCtx.reconcile err pods \"ngrok-proxy-69b849644c-bh4jt\" is forbidden: unable to validate against any pod security policy: [spec.securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used spec.containers[0].hostPort: Invalid value: 6666: Host port 6666 is not allowed to be used. Allowed ports: [] spec.containers[1].hostPort: Invalid value: 7789: Host port 7789 is not allowed to be used. Allowed ports: [] spec.securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used spec.containers[0].hostPort: Invalid value: 6666: Host port 6666 is not allowed to be used. Allowed ports: [] spec.containers[1].hostPort: Invalid value: 7789: Host port 7789 is not allowed to be used. Allowed ports: []]" generation=2 namespace=development resourceVersion=943663544 rollout=ngrok-proxy
time="2022-02-10T04:29:09Z" level=info msg="Reconciliation completed" generation=2 namespace=development resourceVersion=943663544 rollout=ngrok-proxy time_ms=344.57268400000004
time="2022-02-10T04:29:09Z" level=error msg="rollout syncHandler error: pods \"ngrok-proxy-69b849644c-bh4jt\" is forbidden: unable to validate against any pod security policy: [spec.securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used spec.containers[0].hostPort: Invalid value: 6666: Host port 6666 is not allowed to be used. Allowed ports: [] spec.containers[1].hostPort: Invalid value: 7789: Host port 7789 is not allowed to be used. Allowed ports: [] spec.securityContext.hostNetwork: Invalid value: true: Host network is not allowed to be used spec.containers[0].hostPort: Invalid value: 6666: Host port 6666 is not allowed to be used. Allowed ports: [] spec.containers[1].hostPort: Invalid value: 7789: Host port 7789 is not allowed to be used. Allowed ports: []]" namespace=development rollout=ngrok-proxy
time="2022-02-10T04:29:09Z" level=info msg="rollout syncHandler queue retries: 2 : key \"development/ngrok-proxy\"" namespace=development rollout=ngrok-proxy
time="2022-02-10T04:29:09Z" level=info msg="Started syncing rollout" generation=2 namespace=development resourceVersion=943663554 rollout=ngrok-proxy
...
```

**From the ngrok-proxy Rollout status field, we can see a discrepency between replicas/readyReplicas and the messages displayed:**
```
Status:
  HPA Replicas:        4
  Available Replicas:  4
  Blue Green:
  Canary:
    Current Background Analysis Run Status:
      Name:    ngrok-proxy-69b849644c-2
      Status:  Running
  Conditions:
    Last Transition Time:  2022-02-10T03:31:48Z
    Last Update Time:      2022-02-10T03:31:48Z
    Message:               Rollout has minimum availability
    Reason:                AvailableReason
    Status:                True
    Type:                  Available
    Last Transition Time:  2022-02-10T03:56:01Z
    Last Update Time:      2022-02-10T03:56:01Z
    Message:               RolloutCompleted
    Reason:                RolloutCompleted
    Status:                False
    Type:                  Completed
    Last Transition Time:  2022-02-10T04:29:07Z
    Last Update Time:      2022-02-10T04:29:08Z
    Message:               ReplicaSet "ngrok-proxy-69b849644c" is progressing.
    Reason:                ReplicaSetUpdated
    Status:                True
    Type:                  Progressing
    Last Transition Time:  2022-02-10T04:29:08Z
    Last Update Time:      2022-02-10T04:29:08Z
    Message:               Rollout is paused
    Reason:                RolloutPaused
    Status:                False
    Type:                  Paused
  Current Pod Hash:        69b849644c
  Current Step Hash:       6bf54486f
  Current Step Index:      8
  Message:                 old replicas are pending termination
  Observed Generation:     2
  Phase:                   Progressing
  Ready Replicas:          4
  Replicas:                4
  Selector:                app=ngrok-proxy
  Stable RS:               69b849644c
  Updated Replicas:        3
Events:
  Type    Reason                Age                  From                 Message
  ----    ------                ----                 ----                 -------
  Normal  RolloutUpdated        38m                  rollouts-controller  Rollout updated to revision 2
  Normal  NewReplicaSetCreated  37m                  rollouts-controller  Created ReplicaSet ngrok-proxy-69b849644c (revision 2)
  Normal  ScalingReplicaSet     37m                  rollouts-controller  Scaled up ReplicaSet ngrok-proxy-69b849644c (revision 2) from 0 to 1
  Normal  AnalysisRunRunning    37m                  rollouts-controller  Background Analysis Run 'ngrok-proxy-69b849644c-2' Status New: 'Running' Previous: ''
  Normal  RolloutStepCompleted  36m                  rollouts-controller  Rollout step 1/8 completed (setWeight: 20)
  Normal  RolloutStepCompleted  9m6s                 rollouts-controller  Rollout step 2/8 completed (pause: 30m)
  Normal  ScalingReplicaSet     9m6s                 rollouts-controller  Scaled down ReplicaSet ngrok-proxy-75c77696c9 (revision 1) from 3 to 2
  Normal  ScalingReplicaSet     9m5s                 rollouts-controller  Scaled up ReplicaSet ngrok-proxy-69b849644c (revision 2) from 1 to 2
  Normal  RolloutStepCompleted  7m59s                rollouts-controller  Rollout step 3/8 completed (setWeight: 40)
  Normal  RolloutStepCompleted  7m17s                rollouts-controller  Rollout step 4/8 completed (pause: 30m)
  Normal  RolloutStepCompleted  7m17s                rollouts-controller  Rollout step 5/8 completed (setWeight: 60)
  Normal  RolloutStepCompleted  6m44s                rollouts-controller  Rollout step 6/8 completed (pause: 30m)
  Normal  ScalingReplicaSet     6m44s                rollouts-controller  Scaled down ReplicaSet ngrok-proxy-75c77696c9 (revision 1) from 2 to 1
  Normal  ScalingReplicaSet     6m43s                rollouts-controller  Scaled up ReplicaSet ngrok-proxy-69b849644c (revision 2) from 2 to 3
  Normal  RolloutStepCompleted  5m32s                rollouts-controller  Rollout step 7/8 completed (setWeight: 80)
  Normal  RolloutPaused         5m31s (x4 over 36m)  rollouts-controller  Rollout is paused (CanaryPauseStep)
  Normal  RolloutStepCompleted  4m54s                rollouts-controller  Rollout step 8/8 completed (pause: 30m)
  Normal  ScalingReplicaSet     4m54s                rollouts-controller  Scaled down ReplicaSet ngrok-proxy-75c77696c9 (revision 1) from 1 to 0
  Normal  RolloutCompleted      4m53s                rollouts-controller  Rollout completed update to revision 2 (69b849644c): Completed all 8 canary steps
```

Notice that according to the messages (which are correct), one RS is set to 0 and one to 3 - however Replicas and readyReplicas are both 4. From querying kubernetes, we can see that there are in fact only 3 replicas:
```
gavin.kliger@C02CD0T0MD6R:~/universe$ kubectl get pods -n development | grep 'ngrok'
ngrok-proxy-69b849644c-bh4jt                     2/2     Running            0          2m3s
ngrok-proxy-69b849644c-rfxml                     2/2     Running            0          33m
ngrok-proxy-69b849644c-th4mh                     2/2     Running            0          4m25s
```

Additionally, we can see from Kubernetes event logs that the stable pods are all deleted without issue:
```
5m50s       Normal    SuccessfulDelete         replicaset/ngrok-proxy-75c77696c9   Deleted pod: ngrok-proxy-75c77696c9-nwv4h
3m28s       Normal    SuccessfulDelete         replicaset/ngrok-proxy-75c77696c9   Deleted pod: ngrok-proxy-75c77696c9-7rzvz
98s         Normal    SuccessfulDelete         replicaset/ngrok-proxy-75c77696c9   Deleted pod: ngrok-proxy-75c77696c9-d66t5`
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritize the issues with the most 👍.

**Comment by agrawroh** [comment]
@gavin-db Thanks for providing the detailed description and logs!

**Comment by huikang** [comment]
@gavin-db , thanks for reporting. Could you help provide more details regarding the following questions?

1. Are you able to reproduce the error using the latest controller image  or v1.2.0-rc1?
2. what are the canary steps in the rollout spec?
3. What are the steps to reproduce the behavior? (e.g., kubectl apply, rollout update, etc)

**Comment by gavin-db** [comment]
@huikang 
1. I'm not sure - I haven't tried.
2. Note that the exact strategy steps seem irrelevant, as I have reproduced this with two different services (one of which uses more (and longer) pauses.
`        "strategy": {
            "canary": {
                "analysis": {
                    "templates": [
                        {
                            "templateName": "privatelink-proxy-health-checks"
                        }
                    ]
                },
                "canaryMetadata": {
                    "labels": {
                        "hmr_role": "canary"
                    }
                },
                "stableMetadata": {
                    "labels": {
                        "hmr_role": "stable"
                    }
                },
                "steps": [
                    {
                        "setWeight": 20
                    },
                    {
                        "pause": {
                            "duration": "5m"
                        }
                    },
                    {
                        "setWeight": 40
                    },
                    {
                        "pause": {
                            "duration": "5m"
                        }
                    },
                    {
                        "setWeight": 60
                    },
                    {
                        "pause": {
                            "duration": "5m"
                        }
                    },
                    {
                        "setWeight": 80
                    },
                    {
                        "pause": {
                            "duration": "5m"
                        }
                    }
                ]
            }
        },`
3. To reproduce, just kubectl apply a a Rollout manifest with changes that will trigger a release, and then wait. The issue also works if you manually promote after each wait step (to save time). Note that this does not affect all of our services, and so likely has something to do with PSP.

**Comment by agrawroh** [comment]
I took a deeper look at this and don't think it has to do anything with the Argo Rollouts controller. The controller delegate the template(s) verification to the K8S and its K8S which is failing the validation of PSP and we are just propagating that failure back to the user/client.

There are two possible theories:
1. The version of k8s library we are using in Argo v1.2 is newer than the k8s version being used in our control place. Argo Rollouts controller is using the library version **v1.20.4** and our control plane is still on **v1.16** so it's possible that some of the specs got changed and when the controller is trying to validate our templates against these new specs, it's failing.
2. The service pod for which the PSP template validation is failing is running on a privileged node. Argo Rollouts controller is running on a different (unprivileged) node and doesn't have a privileged user so when it's trying to validate the PSP template, it's failing the validation as the pod security policy would block any non-privileged users.

We might have a way to verify **2** by moving the Argo Rollouts controller pod to the same privileged node our service is running on and see if that resolves this issue.
cc @gavin-db
